Add MultiApps Documentation to VTB (#102)
idaholab/virtual_test_bed: doc/content/resources/multiapps_chps/chp_5_multiapps.md

 If a tight coupling approach is desired,  Picard [`fixed point iterations`](https://mooseframework.inl.gov/syntax/Executioner/FixedPointAlgorithms/index.html) may be used by specifying appropriate `Executioner` parameters for the parent application. Picard iterations are nonlinear iterations for tight coupling which repeat the simulation at the same time-step until a convergence criterion is satisfied.  Picard iterations are particularly important if the physics solved in the different apps have very strong nonlinear effects.  

The following two examples describe the use of the `TransientMultiApp`. The definitions of execute_on and positions will be given in the optional parameters.

!listing
[MultiApps]
  [dynamic_executable]
    type = TransientMultiApp
    execute_on = 'TIMESTEP_END'
    input_files = 'pebble.i'
  []
[]

!listing
[MultiApps]
 [sub]
    type = TransientMultiApp
    app_type = BisonApp
    input_files = 'Bison_input.i'
    positions = '0   0   0
                 0.5 0.5 0
                 0.9 0.9 0'
  []
[]

> The minimum time step over the main and all child applications will be used, by default, unless `sub_cycling` is enabled. If sub_cycling is enabled this will allow faster calculations, as the child applications will be able to perform multiple time steps per execution.

Reviewer: ```suggestion
The minimum time step over the main and all child applications will be used, by default, unless `sub_cycling` is enabled. If sub_cycling is enabled this will allow faster calculations, as the parent app will not be forced to use the smaller time step from the child app.
```

### Type of MultiApp

The `type` parameter describes the time/space-based execution of the child application. Various MultiApp `type`s are available to facilitate different possible solution schemes of multi-physics problems. Available `type`s include: [`FullSolveMultiApp`](https://mooseframework.inl.gov/source/multiapps/FullSolveMultiApp.html), [`TransientMultiApp`](https://mooseframework.inl.gov/source/multiapps/TransientMultiApp.html), [`CentroidMultiApp`](https://mooseframework.inl.gov/source/multiapps/CentroidMultiApp.html), [`Stochastic Tools`](https://mooseframework.inl.gov/modules/stochastic_tools/index.html) , and [`LevelSetReinitializationMultiApp`](https://mooseframework.inl.gov/source/multiapps/LevelSetReinitializationMultiApp.html).  

#### FullSolveMultiApp

> The [`FullSolveMultiApp`](https://mooseframework.inl.gov/source/multiapps/FullSolveMultiApp.html) object performs a complete child application simulation during each execution which means that the child applications will be solved from the first-time step to the last time step every time it is called.

Reviewer: ```suggestion
The [`FullSolveMultiApp`](https://mooseframework.inl.gov/source/multiapps/FullSolveMultiApp.html) object performs a complete child application simulation during each execution which means that the child applications will be solved from the first-time step to the last time step every time it is called. This is typically done for steady state or eigenvalue parent-app simulations or for initialization purposes.
```

    app_type = GriffinApp
    input_files = 'Griffin_input.i'
  []
[]

#### MultiApp Classes in `Stochastic Tools App`

The [`Stochastic Tools`](https://mooseframework.inl.gov/modules/stochastic_tools/index.html) module is a toolbox designed for performing stochastic analysis for MOOSE-based applications. It can be used as a parent application to statistically control some key parameters in other MOOSE applications as its child applications.

 The [`Stochastic Tools`](https://mooseframework.inl.gov/modules/stochastic_tools/index.html) apps include [`SamplerFullSolveMultiApp`](https://mooseframework.inl.gov/source/multiapps/SamplerFullSolveMultiApp.html) , [`SamplerTransientMultiApp `](https://mooseframework.inl.gov/source/multiapps/SamplerTransientMultiApp.html) , and [`PODFullSolveMultiApp`](https://mooseframework.inl.gov/source/multiapps/PODFullSolveMultiApp.html) . [`SamplerFullSolveMultiApp`](https://mooseframework.inl.gov/source/multiapps/SamplerFullSolveMultiApp.html)  and [`SamplerTransientMultiApp `](https://mooseframework.inl.gov/source/multiapps/SamplerTransientMultiApp.html) create full solve type, and transient type child application, respectively.

> #### MultiApp Classes in `Level Set App`
> 
> The [`LevelSetReinitializationMultiApp`](https://mooseframework.inl.gov/source/multiapps/LevelSetReinitializationMultiApp.html) method is used for front tracking problems, and its mathematical formulations represents the structural solid-void boundary.

Reviewer: ```suggestion
```
too specific imo

    app_type  = BisonApp
    positions =   '0.01285  0.0       0
                   0.01285  0.0608    0
                   0.01285  0. 1216   0
                   0.01285  0.1824    0'
    input_files = name.i
    execute_on = 'timestep_end'
  []
[]

### Optional Parameters

#### Timing of Execution

The `execute_on` parameter specifies when a child application is to be executed, in relation to the parent application timesteps. The most common options are to execute the child application at the beginning of the time step, at the end of the time step, or at the initial condition of the entire simulation. By default, child application execution will occur at the beginning of time-step.  In the previous example, at every time step, the main application will be executed first, then the child applications will be executed at the end of the time step. Some common options of `execute_on` are summarized below:

  1. `TIMESTEP_BEGIN`: Launch child application at the beginning of the time step

  2. `TIMESTEP_END`: Launch child application at the end of the time step

  3. `INITIAL`: Launch child application only at the initial condition to set up the initial conditions for the solve

>   4. `NONLINEAR`: This option has limited applications and is used to inject the coupling information into the middle of the Newton solve which will slow down the newton convergence but could speed up the conversion of the entire system.  

Reviewer: ```suggestion
  4. `NONLINEAR`: This option has limited applications and is used to inject the coupling information into the middle of the Newton solve. It will likely slow down the newton convergence, as the contribution to the Jacobian are not considered, but could speed up the convergence of the entire system.  
```

#### Finer Timesteps in Child Applications

Applications may use different time step sizes by allowing `subcycling`. The subcycling option is useful for allowing different timescales. For instance, a relatively slow physics application may use bigger time steps (e.g. 10 sec) whereas a rapidly changing physics may use relatively small-time step (e.g. 0.1 sec). Thus, the first app takes one time-step of 10 sec while the other app takes 100 time-steps of 0.1 sec to reach the same end time as the first application.

!media media/resources/advsubcycling.png
      style=display: block;margin-left:auto;margin-right:auto;width:80%;
      id= subcycling
      caption=Schematic drawing of the sub-cycling concept.

>  [subcycling] illustrates a transient Multiphysics simulation in which Griffin, Bison, and SAM are coupled at different timescales using the subcycling option. When performing sub-cycling, transferred auxiliary variables on child applications may be interpolated between the start time and the end time of the main app with the `interpolate_transfers` parameter. When using the TransientMultiApp type, the `allow_sub_cycling` parameter can be used to invoke subcycling.

Reviewer: ```suggestion
 [subcycling] illustrates a transient Multiphysics simulation in which Griffin, Bison, and SAM are coupled at different timescales using the subcycling option. When performing sub-cycling, transferred auxiliary variables on child applications may be interpolated between the start time and the end time of the main app with the `interpolate_transfers` parameter. When using the TransientMultiApp type, the `sub_cycling` parameter can be used to invoke subcycling.
```

#### Locations of Child Applications

> Child applications can be launched at various spatial positions of the parent application geometry by using the 'positions' input parameter. Positions are simply a list of 3D coordinates that describe the offset of the child application's origin from the parent application's origin. If no positions are provided, the child application will be launched at position is `0 0 0` within the parent application. Notice that in the example below, the number of child applications to be invoked is not explicitly defined but rather is inferred from the number of positives given. The example shows five different positions, thus, five independent versions of the child application `BisonApp` will be created and each of them will be associated with a different position within the parent application domain. The user may also provide a position file instead of explicitly providing the positions in the MultiApp block; this feature is useful when the user wants to invoke thousands of child application as the separate input file can contain all this information without cluttering the main input file. A common use of positions in nuclear reactor analysis is the launching of individual fuel rod (BISON) simulations at various fuel rods within a reactor assembly or core.  

Reviewer: ```suggestion
Child applications can be launched at various spatial positions of the parent application geometry by using the 'positions' input parameter. Positions are simply a list of 3D coordinates that describe the offset of the child application's origin from the parent application's origin. If no positions are provided, the child application will be launched at the position `0 0 0` within the parent application. Notice that in the example below, the number of child applications to be invoked is not explicitly defined but rather is inferred from the number of positions given. The example shows five different positions, thus, five independent versions of the child application `BisonApp` will be created and each of them will be associated with a different position within the parent application domain. The user may also provide a position file instead of explicitly providing the positions in the MultiApp block; this feature is useful when the user wants to invoke thousands of child application as the separate input file can contain all this information without cluttering the main input file. A common use of positions in nuclear reactor analysis is the launching of individual fuel rod (BISON) simulations at various fuel rods within a reactor assembly or core.  
```

         block=MultiApps

The list and location of each heat-pipe are provided in +hp_centers.txt+ as follows:

!listing mrad/steady/hp_centers.txt        

#### Parallel Options

 When the mesh requires large memory usage, it is useful to use [`Distributed Mesh`](https://mooseframework.inl.gov/syntax/Mesh/) by setting `parallel_type` as `distributed` in the `Mesh` block so that the different portions can be stored in different processors.

 It is also useful to assign a few processors for small solves using `max_procs_per_app` which is the maximum number of processors to give to each App. However, for large solves, the user can assign the minimum number of processors to give to each App using `min_procs_per_app`.

#### Other Options

 For the MOOSE Apps: `FullSolveMultiApp`, `TransientMultiApp`, and `CentroidMultiApp`. There are various common optional parameters such as `app_type`, `bounding_box_inflation`, `bounding_box_padding`,  `catch_up`, `cli_args`, `cli_args_files` , `clone_master_mesh` , `detect_steady_state`,  `global_time_offset` , `keep_solution_during_restore`, `library_name`, `library_path`,`output_in_position`, and `execute_on`.

>  Other parameters are allowed only for specific MultiApps type. For example, `positions` is an optional parameter for FullSolveMultiApp and TransientMultiApp, but it is not a paramter for CentroidMultiApp. Generally, these optional parameters were developped to facilitate flexiabile multi-physics simulations.  The `no_backup_and_restore` optional parameter is useful when doing steady-state Picard iterations where we want to use the solution of previous Picard iteration as the initial guess of the current Picard iteration.

Reviewer: ```suggestion
 Other parameters are allowed only for specific MultiApps type. For example, `positions` is an optional parameter for FullSolveMultiApp and TransientMultiApp, but it is not a parameter for CentroidMultiApp. Generally, these optional parameters were developed to facilitate flexible multi-physics simulations.  The `no_backup_and_restore` optional parameter is useful when doing steady-state Picard iterations where we want to use the solution of previous Picard iteration as the initial guess of the current Picard iteration.
```
